File: Task Tracker (GO)/manager.go
package main

import (
	"errors"
	"time"
)

type TaskManagerImpl struct {
	storage Storage
	tasks   []Task
	nextID  int
}

func NewTaskManager(storage Storage) (*TaskManagerImpl, error) {
	tm := &TaskManagerImpl{
		storage: storage,
		nextID:  1,
	}

	tasks, err := storage.Load()
	if err != nil {
		return nil, err
	}

	tm.tasks = tasks
	if len(tasks) > 0 {
		tm.nextID = tasks[len(tasks)-1].ID + 1
	}

	return tm, nil
}

func (tm *TaskManagerImpl) Add(description string) error {
	now := time.Now()
	task := Task{
		ID:          tm.nextID,
		Description: description,
		Status:      Todo,
		CreatedAt:   now,
		UpdatedAt:   now,
		CompletedAt: time.Time{},
	}

	tm.tasks = append(tm.tasks, task)
	tm.nextID++
	return tm.storage.Save(tm.tasks)
}

func (tm *TaskManagerImpl) UpdateStatus(id int, status Status) error {
	now := time.Now()
	for i := range tm.tasks {
		if tm.tasks[i].ID == id {
			oldStatus := tm.tasks[i].Status
			tm.tasks[i].Status = status
			tm.tasks[i].UpdatedAt = now

			if status == Done && oldStatus != Done {
				tm.tasks[i].CompletedAt = now
			}

			if status != Done && oldStatus == Done {
				tm.tasks[i].CompletedAt = time.Time{}
			}

			return tm.storage.Save(tm.tasks)
		}
	}
	return errors.New("tarea no encontrada")
}

func (tm *TaskManagerImpl) Delete(id int) error {
	for i, task := range tm.tasks {
		if task.ID == id {
			tm.tasks = append(tm.tasks[:i], tm.tasks[i+1:]...)
			return tm.storage.Save(tm.tasks)
		}
	}
	return errors.New("tarea no encontrada")
}

func (tm *TaskManagerImpl) ListAll() ([]Task, error) {
	return tm.tasks, nil
}

func (tm *TaskManagerImpl) ListByStatus(status Status) ([]Task, error) {
	var result []Task
	for _, task := range tm.tasks {
		if task.Status == status {
			result = append(result, task)
		}
	}
	return result, nil
}

func (tm *TaskManagerImpl) GetTaskCount() int {
	return len(tm.tasks)
}

File: TaskTrackerRust/task_tracker_rust/src/manager.rs
use crate::storage::Storage;
use crate::task::{Status, Task};
use std::io;

pub trait TaskManager {
    fn add(&mut self, description: String) -> io::Result<()>;
    fn update_status(&mut self, id: i32, status: Status) -> io::Result<()>;
    fn delete(&mut self, id: i32) -> io::Result<()>;
    fn list_all(&self) -> Vec<Task>;
    fn list_by_status(&self, status: Status) -> Vec<Task>;
    fn get_task_count(&self) -> usize;
}

pub struct TaskManagerImpl<'a> {
    pub storage: &'a dyn Storage,
    pub tasks: Vec<Task>,
    next_id: i32,
}

impl<'a> TaskManagerImpl<'a> {
    pub fn new(storage: &'a dyn Storage) -> io::Result<Self> {
        let tasks = storage.load()?;
        let next_id = tasks.iter().map(|t| t.id).max().unwrap_or(0) + 1;
        Ok(TaskManagerImpl {
            storage,
            tasks,
            next_id,
        })
    }

    fn persist(&self) -> io::Result<()> {
        self.storage.save(&self.tasks)
    }
}

impl<'a> TaskManager for TaskManagerImpl<'a> {
    fn add(&mut self, description: String) -> io::Result<()> {
        let now = Task::now_timestamp();
        let task = Task {
            id: self.next_id,
            description,
            status: Status::Todo,
            created_at: now,
            updated_at: now,
            completed_at: None,
        };
        self.tasks.push(task);
        self.next_id += 1;
        self.persist()
    }

    fn update_status(&mut self, id: i32, status: Status) -> io::Result<()> {
        let now = Task::now_timestamp();
        for t in self.tasks.iter_mut() {
            if t.id == id {
                let was_done = t.status == Status::Done;
                t.status = status.clone();
                t.updated_at = now;
                if status == Status::Done && !was_done {
                    t.completed_at = Some(now);
                } else if status != Status::Done {
                    t.completed_at = None;
                }
                return self.persist();
            }
        }
        Err(io::Error::new(io::ErrorKind::NotFound, "Tarea no encontrada"))
    }

    fn delete(&mut self, id: i32) -> io::Result<()> {
        if let Some(pos) = self.tasks.iter().position(|t| t.id == id) {
            self.tasks.remove(pos);
            return self.persist();
        }
        Err(io::Error::new(io::ErrorKind::NotFound, "Tarea no encontrada"))
    }

    fn list_all(&self) -> Vec<Task> {
        self.tasks.clone()
    }

    fn list_by_status(&self, status: Status) -> Vec<Task> {
        self.tasks
            .iter()
            .filter(|t| t.status == status)
            .cloned()
            .collect()
    }

    fn get_task_count(&self) -> usize {
        self.tasks.len()
    }
}
